### ISSUE TITLE
object generic properties - behavior is not quite rational

### ISSUE THREAD

**Issue by craigphicks** [microsoft/TypeScript#42674]
# Bug Report

### 🔎 Search Terms

Object generic properties

### 🕗 Version & Regression Information

I noticed this behavior as I was attempting to map JSON schema design to Typescript behavior.  (Extending JSON Schema to internal use cases).

- I was unable to test this on prior versions because unlikely to be different

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/dev/bug-workbench/?#code/FAehAIBUAsEsGdwCMCuBzcAKA9gJ3AIYB22AttgDbYqJICm0BAbrHgJTi50AOeALuADG2IoLq4iiAETYkAKzqCBaOkXGxB4brmzdxfWHXhSAdKAgxxdQl3B8A7tnABrWEQAmibADMXdAJ6EFFT2dO7gbuAqargaWjp6uAZGAFzgANrOASnwfLFEaAC6ADQZWf4pRCik9LiF5uAAYnjg8LCk3BQasHyBIhSBggTwRuD2PdDUAl184vHY+oaIBLbuCII0I+EAtBECCK0o3t7dqnxmYFDQ1t6UIW4YQyPLtj4RRLNcuSnADQDCw2sAEY0sBenpwAB5eRA8AAXnAAG9gJlsq08g9CmkqjVxABuYAAXwJ-0B4AATD9wdZoXJyfCkSjymlcvkitjqrUCaiKjjali+fiiSTLgCRuAAMxpTAEcAAcjUTHEctadAEuxI8yQFDopDGEzsjD4bDB-ghtIlDORPJyGIKAroStw3OZgrqlU5c3A3suJAEwlE4jU4XGfGghoO4h0uGFv1FZIALGkAISprCCRgFazaXTYeAECjIOghE3UqHyBNWpnZN0Cz3O2Ok8UAVhTacwGeIKnmvHzhfoJdN5vkzarNtrlUdQptrMxHtxDeJvybo1hxHC9JW1lg7jOGgLhA8rXanUCBZCQULbm4KD4pQAjihsLNwi0UERH8+wqV1+BSARuENPNrCYAsUFGYZwFuYJsHGApwB1NACEEQIACkAGVOBQHV4AuCAAGkAnAAARMgCDcNJwE-F9SnfaiwnhAA+S9CEEMRuBfBpCMCABZADuAeNJbnwLgWDoexwAab1vXgFJ6PCOFmPgUppNUtT1I0zSsB3PdenAIFtlhf9uAEgoTUuVTZLop8XyYiJ4AAOQIBzMAc+tMHgNgOAAflacA0jchcPI4b1MH-IhAj4JwjP4h5wCmNpd3i3x828bcPjoFRcE8riiNwLs6Eo-Ks3sljZwKRBMDDYhnEQKKIl3D4ekCYzTLQE0V0lSiwwOEZ9kQXUOLPI9kLYng+EQTVylw710IIUhrEgxEZztdlFXEQkghEOgzE6pMrgOJ5t0QeAUBM4Dwnq0bxvAN1liPLgFCUW6RG2O6TG9XLAlI-8KMkiyZJSayvwU5jkwQJyiGCqTAfk6TFPAcHHOIVz3M8jgADIMfAQLamCkwovQ1bMDYOEyfgL7OAKn4Ab83DzkRiGUdx8Rgsx7GWdwfHCeJ0nybjCxrlY9i+GIMRkqomyGPKtBEAOP1wBoOZvFiVR3AGH4A1yeKKwIFJaUrBFETlIFyQlBNmwANgAdgADgATgABlN82retuU0lhTbwF9Z9hfGhj6CGJXvU57nsCJtkSeTcnQpNs2LZth3nYTt2VRjhF49dpOnZdxPbblE0tYEWQ5ATJB9YrK0s8Tu3c9Ty2C898Bvcua6OMDxQCBDtSw88gmI95sm4UQLAa7d5O86tgv4TJ+Up5zlPs4Ljr4xbNIYEOslerO3sGKusaOJeog3vrQ9wkexQBHeymfvIohQVpqyPyl0H7Mh6Gn7k1-4eYpmob7l5cAAAfYBOM0ZsAHpHB40d+YNEsK0AgaV3izCyi8RawROB0CerZVAAh3ypXSqg8Q6DWhkDVO0CCh9bK-ngOQgwC1ECXyUGED6mAEEKmfCqN4R1wAJiLiIbWpdmx61pKOI249F4L3ds3VuEB264K7j3QBUDeYZ1HpgSRdcl613TnPLRk8G4r2AMXHWchmwVzEdXaRhjl4e30i3H0EAFYKM7sHcUvcIGqKjnzEeccbH1zsbPTOASdHT0LgLcAYprCjl6gQfwcU7qHAzIQRABjAm6JsKoOUAguDnkCEkoO3dxRhiIu4EQOS-wATsE4UppARgUCVIgd8u58AYSwjhPC0kACSAhylGCIJUtKxZOCwDQNAO859SqyiqAgMWNwWhuE+N4ZCDxdpgA2ZsrZ2ydm7L2fsg5hyjmHIaAABQSHmAsiBhJ-mwGsE4ndGAsBaNsV5byGgOUhJAAAomkHqiAcy9gPP0yaftOwlVKVUkycUWlzHaW8cpv0iB-l1LUZSMNVI9HikQAYWDZxKDqkLRF993iHgiB0HQSoQzxMidEvhG84CIF4Tvc6Wwan+yPkk38zDr6vXep9S43ESJkT+hi2S8laIvxBt6BGnMTAIDmmlbpGUsqoyCujSmxUVA0wgNJc8dMCYfTlQqpBdBlXEK5oA1eEA6WtgOky7eJ1d4XXZQo4+p8Fz3Qvtgq+t16wzX+gRIid9RVfwlYrKVtk7JMxcjlQVeVqYYv1SdQ179nLQwaMq1okwkilDyHQAgAg4a-mBrZHSTUhjBH8D+I8NzygctFoGDpoxdwnDUPKO6KpIKQsISgzKJCa3hE4SfTtqT3U3yAA)

### 💻 Code
```

// This bug (or anomolous behavior) report concerns "object generic properties".
// There are two kinds of key allowed in generic properties: [key:string], [key:number]
// For simplicity only cases without liter proerties are discussed - it is sufficient.
// The following cases are of interest:

// Case 1: 
type Obj1 = {
[key: string]: number;
};

// Case 2:
type Obj2 = {
[key: string]: number;
[key:number]:number;
};

// Case 3: (a 'never' set - no problem with that)
type Obj3 = {
[key:string]:never;
[key:number]:number    // not concerned with this error
};

// Case 4: !!! (change proposal below)
type Obj4 = {
[key:number]:number;
};

// Case 5: !!! (change proposal below)
type Obj5 = {
[key:number]:never;
[key:string]:number;
};


// Cases 1 and 2 are identical and simply allow all input, quoted or unquoted, and map those values as following legacy JS rules.
// Key Domain:  quoted, unquoted => all accepted
// Key Mapping: for review 
//     s:quoted => s,                                   (identity 1-1 mapping)
//     s:unquoted => isNaN(Number(s)) ? s : Number(s)   (many to 1 mapping outside of safe integers)
// Key range:  range is all strings (thanks to identity mapping)

// Case 3:  this set is empty and acccepts no keys.  Same as {[key:string]:never} alone.

// Case 4: This case is supposed to accept numbers and reject non-numbers.  
// Key Domain: 
//    s:unquoted => !isNan(s)
//    s:quoted   => !isNan(Number(s)) && Number(s).toString()===s
// Key range:
//    s s.t. !isNan(Number(s)) && Number(s).toString()===s

// The acceptance of quoted strings is not user friendly:
const obj4a:Obj4 = {'12345678901234567': 1 }  // not accepted because  Number(s).toString()!==s  ('12345678901234567' !== '12345678901234568')
const obj4b:Obj4 = {'12345678901234568': 1 }  // accepted because      Number(s).toString()===s  ('12345678901234568' === '12345678901234568')

// Case 5: This case is supposed to accept non-number and reject numbers.
// Key Domain: 
//    s:unquoted => isNan(s)
//    s:quoted   => isNan(Number(s)) || Number(s).toString()!==s

// The safe integers are all rejected but unsafe integers are sometimes accepted and sometimes rejected. (The 'not' of case 4)
const obj5a:Obj5 = {'12345678901234567': 1 }  // accepted because  Number(s).toString()!==s  ('12345678901234567' !== '12345678901234568')
const obj5b:Obj5 = {'12345678901234568': 1 }  // not accepted because      Number(s).toString()===s  ('12345678901234568' === '12345678901234568')

// Case 5 is saying numbers such as '12345678901234567' aren't really numbers because they don't map to themselves under JS rules.
//   It doesn't feel right, and is a nuisance for interfacing.

//////////////////////////////////////////////////////////////////////////////////////////////////////
// Proposals for modified behavior ------
// NOTE: this proposal does not change the mapping under JS of domain members,
//       it only restricts the domain in an improved way

// Case 4: This case is supposed to accept numbers and reject non-numbers.  
// Key Domain: 
//    s:quoted, unquoted  => Number.isSafeInteger(Number(s))
// Key range:
//    all s s.t. Number.isSafeInteger(Number(s))

// Case 5: This case is supposed to accept non-numbers and reject numbers.  
// Key Domain: 
//    s:quoted, unquoted  => isNaN(s)
// Key range:
//    all s s.t. isNaN(s)

// In short, treat quoted and unquoted identically, and for key acceptance rules define 'numbers' as the safe integers, and 'non-numbers' as non-numbers.

```

#### Case 4: 
```
  type Obj4 = {
  [key:number]:number;
};
```
The acceptance of quoted strings is not user friendly:

 - This is not accepted because  `Number(s).toString()!==s  ('12345678901234567' !== '12345678901234568')`
    -  `const obj4a:Obj4 = {'12345678901234567': 1 }`  
 - while this is accepted because   `Number(s).toString()===s  ('12345678901234568' === '12345678901234568')`
    -  `const obj4b:Obj4 = {'12345678901234568': 1 }`  

#### Case 5:
```
type Obj5 = {
  [key:number]:never;
  [key:string]:number;
};
```

  - The safe integers are all rejected but unsafe integers are sometimes accepted and sometimes rejected. (The 'not' of case 4)
  - This is accepted because  `Number(s).toString()!==s  ('12345678901234567' !== '12345678901234568')`
      `const obj5a:Obj5 = {'12345678901234567': 1 }`  
  - while this is not accepted because  `Number(s).toString()===s  ('12345678901234568' === '12345678901234568')`
      `const obj5b:Obj5 = {'12345678901234568': 1 }`  
  - Case 5 is saying numbers such as '12345678901234567' aren't really numbers because they don't map to themselves under JS rules.  It doesn't feel right, and is a nuisance for interfacing.
  

### 🙂 Expected behavior

Proposals for modified behavior 
------
*NOTE*: this proposal does not change the mapping under JS of domain members, it only restricts the domain in an improved way.

#### Case 4: 
```
type Obj4 = {
  [key:number]:number;
};
```
This case is supposed to accept numbers and reject non-numbers.  

 -  Key Domain: 
    - `s:quoted, unquoted  => Number.isSafeInteger(Number(s))`
 - Key range:
    - `all s s.t. Number.isSafeInteger(Number(s))`

#### Case 5: 
```
type Obj5 = {
  [key:number]:never;
  [key:string]:number;
};
```
This case is supposed to accept non-numbers and reject numbers.  
 - Key Domain: 
   - `s:quoted, unquoted  => isNaN(s)`
 - Key range:
   - `all s s.t. isNaN(s)`

*In short, treat quoted and unquoted identically, and for key acceptance rules define 'numbers' as the safe integers, and 'non-numbers' as non-numbers.*

**Comment by craigphicks** [comment]
From the POV of the range of literal inputs (quoted|unquoted) under the JS transformation, the existing behavior is rational.
  - The allowable keys of `{[key:number]:T}` is the image of all keys settable with someobj[num] where num is a number.
  - The allowable keys of `{[key:string]:T,[key:number]:T}` is the exact inverse of that image.

The existing behavior might have some strange behavior for a user entering strings of digits, and it might not be that useful, but it does have a solid rationale.  So I will close this issue.
